crypto/s2n_pkey.c

```diff
@@ -0,0 +1,136 @@
+#include <openssl/evp.h>
+#include <openssl/x509.h>
+
+#include "error/s2n_errno.h"
+
+#include "crypto/s2n_pkey.h"
+
+#include "utils/s2n_safety.h"
+
+int s2n_asn1der_to_private_key(struct s2n_pkey *priv_key, struct s2n_blob *asn1der)
+{
+    int ret;
+    uint8_t *key_to_parse = asn1der->data;
+
+    /* Detect key type */
+    EVP_PKEY *pkey = d2i_AutoPrivateKey(NULL, (const unsigned char **)(void *)&key_to_parse, asn1der->size);
+    if (pkey == NULL) {
+        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
+    }
+    
+    /* If key parsing is successful, d2i_AutoPrivateKey increments *key_to_parse to the byte following the parsed data */
+    uint32_t parsed_len = key_to_parse - asn1der->data;
+    if (parsed_len != asn1der->size) {
+        EVP_PKEY_free(pkey);
+        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
+    }
+
+    int type = EVP_PKEY_type(EVP_PKEY_id(pkey));
+    
+    /* Initialize s2n_pkey according to key type */
+    switch (type) {
+    case EVP_PKEY_RSA:
+        ret = s2n_pkey_to_rsa_private_key(&priv_key->key.rsa_key, pkey);
+        priv_key->ctx = &rsa_key_ctx;
+        break;
+    case EVP_PKEY_EC:
+        ret = s2n_pkey_to_ecdsa_private_key(&priv_key->key.ecdsa_key, pkey);
```

Comment:
The functions were used as helper-ish functions to convert between specific types (s2n_rsa_public_key, s2n_ecdsa_public_key) and don't take the same parameters. I tried to keep the s2n_pkey functions more scoped to what you can conceptually do with a public/private key.